client/rpc/debug/src/lib.rs

```diff
@@ -57,49 +96,89 @@ where
 	C::Api: DebugRuntimeApi<B>,
 	C::Api: EthereumRuntimeRPCApi<B>,
 {
-	fn trace_transaction(
-		&self,
+	pub fn task(
+		client: Arc<C>,
+		backend: Arc<BE>,
+		frontier_backend: Arc<fc_db::Backend<B>>,
+		permit_pool: Arc<Semaphore>,
+	) -> (impl Future<Output = ()>, DebugRequester) {
+		let (tx, mut rx): (DebugRequester, _) =
+			sp_utils::mpsc::tracing_unbounded("debug-requester");
+
+		let fut = async move {
+			loop {
+				if let Some(((transaction_hash, params), response_tx)) = rx.next().await {
+					let client = client.clone();
+					let backend = backend.clone();
+					let frontier_backend = frontier_backend.clone();
+					let permit_pool = permit_pool.clone();
+					// Spawn the long running task in another thread.
+					let res = tokio::task::spawn_blocking(move || {
+						return Self::handle_request(
+							client.clone(),
+							backend.clone(),
+							frontier_backend.clone(),
+							transaction_hash,
+							params,
+						);
+					});
+
+					tokio::task::spawn(async move {
+						let _permit = permit_pool.acquire();
```

Comment:
Here the permit is only taken to await on the result of the blocking task and send the response, but **not on spawning the blocking task**, which is what this permit is supposed to control.
You should move your `spawn_blocking` call just after acquiring the permit.

Reply:
Thought `spawn_blocking` was evaluated on await. Thank you! Will fix.